README
The code is organized into three main parts: processing, cutting, and plotting (which are sorted into directories). Each directory contains a one or more executable scripts (process.py, cut.py, comparison.py, and interpolation.py), as well as modules contain the functions used within the executables.

The input and output files for the executable scripts are provided as command line arguments, so it should be easy to automate workflows using shell scripts (several examples are provided). For an explanation of the command line usage, call the script with the -h flag. For example,

$ python process.py -h

To view the documentation for a file, call the pydoc command on it. Eg.

$ pydoc geometry.py
... lots of wonderful documentation

What follows is a general overview of the main executable files.

Processing: Before we do the main cuts on the files, some basic filtering, reconstruction, and calculations must be done. process.py takes an I3 file with its GCD, applies this processing, and writes the output to another I3 file. Here is a general overview of the script:

  o Get the command line arguments and open the files

  o Frame Filtering

    Some basic frame filters (these are held constant). They are:

    in_ice - Check that sub_event_stream == 'in_ice'
    min_bias - Check in FilterMinBias_11 that condition_passed and prescale_passed are both true
    SMT8 - Make sure that the length of TWOfflinePulsesHLC is >= 8
    MPEFit - Check that the fit_status of MPEFit is OK, and that 40 < zenith < 70
    TriggerCheck_12 and InIceSMTTriggered - Add the TriggerCheck_12 module, which calculates InIceSMTTriggered, and check that InIceSMTTriggered is true.

  o General

    General functions:

    direct_hits, hit_multiplicity, hit_statistics - Add these segments to calculate common cut variables

    count_hits - Iterate over the pulse series specified in options['pulse_series'], and calculate ICAnalysisHits and DCAnalysisHits. ICAnalysisHits is the number of hits in the IC analysis region (the inner two hexagons, excluding Deep Core and string 36), and DCAnalysisHits is the number of hits in the Deep Core region, excluding strings 79 and 80.

    reco_endpoint - Calculate the reconstructed endpoint of the event (using the provided fit).

    move_cut_variables - Move the NDirDoms cut value calculated by direct_hits, rlogl from the provided fit parameters, and the z coordinate of the reconstructed endpoint into the top level of the frame. This is done so the cut.py script can access these values directly.

  o Geoanalysis

    These functions look at the geometry of events relative to the detector. The heavy lifting is done in geometry.py, which contains very general functions for simple problems in 2D computational geometry. The functions in geoanalysis mostly translate the data in the frames into something the geometry functions can work with.

    calc_dist_to_border - Calculate the shortest distance of the reconstructed event endpoint to the detector border. Endpoints outside the detector get negative distances, and events inside positive.

  o Domanalysis

    om_partition -

    dom_data -

  o Write the output to an I3 file

process.sh is an example shell script for running cut.py


Cutting: Except for a few basic cuts (min_bias, SMT8, etc.) done in the processing file, the majority of cuts are done here. In the cutting script, an arbitrary number of processed I3 files are provided as input. The specified cuts are then applied, and the data is then written out to an HDF5 file for plotting.

  o Get the command line arguments and open the files

  o make_event_cuts - Remove the frames that do not pass all the event cuts. These cuts are defined within the 'event_cuts' dictionary. It has a simple format. The keys are the names of keys in the frame, and the values define the cut to make. Ex. event_cuts['NDirDoms'] = ('greater than', 5) means only keep frames with an NDirDoms greater than 5. Easy.

  o make_dom_cuts - Now that we have filtered out the frames that do not pass the event cuts, we can do the dom cuts. Each frame has several I3VectorDoubles that contain the data for all the doms that The cuts on the dom data are defined similarly in the 'dom_cuts' dictionary. Once the cuts have been made, the data is split apart into data for IC doms and data for DC doms.

  o Write out data to an HDF5 file. To keep the file size manageable, only user selected keys are written to the file; these are defined in the 'event_keys' and 'dom_keys' lists.

  o write_cut_metadata - Write the cut dictionaries to the HDF5 as metadata (so we can retrieve them later to see what cuts were made).

cut.sh is an example script for running cut.py


Plotting: interpolation.py creates the final plot used to derive the in ice DOM efficiency. To use this script, you need several simulated datasets of various DOM efficiencies, as well as an experimental datafile. The idea is that the charges are placed into bins based on the corresponding reco_distances (0-20 m, 20-40 m, etc.). This is done for each dataset, and then the averaged charges for each bin are scaled down by the corresponding average charge for ______. The scaled average charges in the 20-40 m, 40-60 m, and 60-80 m bins are averaged. This charge is plotted on the y-intercept.

  o Get the command line arguments

  o Specify all the experimental errors


Global Dependencies: (for all scripts)
* Python 2.7 or 3.2+
* Numpy 1.7+

Processing Dependencies:
* IceTray

Cutting Dependencies:
* IceTray
* HDF5 1.8.11+
* Pytables 3.0+

Plotting Dependencies:
* HDF5 1.8.11+
* Pytables 3.0+
* Matplotlib 1.3+


changes.txt
Changes from IC79 to IC86

o border doms updated for IC86 in dist_to_border
o jeb-filter-2011 -> jeb-filter-2012
o TriggerCheck_11 -> TriggerCheck_12
o removed BoundsE, BoundsL, and StepL in I3SimpleParametrizationFactory
o replace I3CutsModule (deprecated) with CommonVariables
o nullsplit -> in_ice
o FilterMinBias_10 -> FilterMinBias_11
o MPEFit_SLC -> MPEFit
o I3Position.calc_distance -> vector subtraction and then magnitude of difference


process/process.py
#!/usr/bin/env python

"""
Process the I3 files.
"""

from __future__ import print_function, division  # 2to3

import argparse

from icecube import dataio, icetray, gulliver, simclasses, dataclasses, photonics_service, phys_services
from icecube.common_variables import direct_hits, hit_multiplicity, hit_statistics
from I3Tray import I3Tray, I3Units, load

from filters import in_ice, min_bias, SMT8, MPEFit, InIceSMTTriggered
from general import get_truth_muon, truth_endpoint, count_hits, reco_endpoint, move_cut_variables
from geoanalysis import calc_dist_to_border
from domanalysis import om_partition, dom_data

load('libipdf')
load('libgulliver')
load('libgulliver-modules')
load('liblilliput')
load('libstatic-twc')
load('libjeb-filter-2012')


def main():

    parser = argparse.ArgumentParser(description='script for proccessing I3 files')
    parser.add_argument('gcd', help='GCD file for the data')
    parser.add_argument('data', help='data file for processing')
    parser.add_argument('ofile', help='name of output file')
    parser.add_argument('-s', '--sim', help='turn on extra processing for sim files',
                        action='store_true')
    args = parser.parse_args()

    # Don't touch
    options = {}
    options['pulses_name'] = 'TWSRTOfflinePulses'
    options['max_dist'] = 140
    options['partitions'] = 5

    tray = I3Tray()

    # Read the files.
    tray.AddModule('I3Reader', 'I3Reader',
                   Filenamelist=[args.gcd, args.data])

    # Filters

    # Filter the ones with sub_event_stream == in_ice
    tray.AddModule(in_ice, 'in_ice')

    # Check in FilterMinBias_11 that condition_passed and prescale_passed are both true
    tray.AddModule(min_bias, 'min_bias')

    # Make sure that the length of TWOfflinePulsesHLC is >= 8
    tray.AddModule(SMT8, 'SMT8')

    # Check that the fit_status of MPEFit is OK, and that 40 < zenith < 70
    tray.AddModule(MPEFit, 'MPEFit')

    # Trigger check
    # jeb-filter-2012
    tray.AddModule('TriggerCheck_12', 'TriggerCheck_12',
                   I3TriggerHierarchy='I3TriggerHierarchy',
                   InIceSMTFlag='InIceSMTTriggered',
                   IceTopSMTFlag='IceTopSMTTriggered',
                   InIceStringFlag='InIceStringTriggered',
                   PhysMinBiasFlag='PhysMinBiasTriggered',
                   PhysMinBiasConfigID=106,
                   DeepCoreSMTFlag='DeepCoreSMTTriggered',
                   DeepCoreSMTConfigID=1010)

    # Check that InIceSMTTriggered is true.
    tray.AddModule(InIceSMTTriggered, 'InIceSMTTriggered')

    # Endpoint

    # Add the reconstructed event endpoint to the frame.
    tray.AddModule(reco_endpoint, 'reco_endpoint',
                   endpoint_fit='FiniteRecoFit')

    # Domanalysis

    # Recalculate recos on subset of Doms (above dust layer)
    # lilliput
    tray.AddService('I3SimpleParametrizationFactory', 'SimpleTrack',
                    StepX=20 * I3Units.m,                              # Set to 1/50 the size of the detector
                    StepY=20 * I3Units.m,                              # Set to 1/50 the size of the detector
                    StepZ=20 * I3Units.m,                              # Set to 1/50 the size of the detector
                    StepZenith=0.1 * I3Units.radian,                   # Set to 1/30 the size of the detector
                    StepAzimuth=0.2 * I3Units.radian,                  # Set to 1/30 the size of the detector
                    StepLinE=0,                                        # Default
                    StepLogE=0,                                        # Default
                    StepT=0,                                           # Default
                    BoundsAzimuth=[0, 0],                              # Default
                    BoundsZenith=[0, 0],                               # Default
                    BoundsT=[0, 0],                                    # Default
                    BoundsX=[-2000 * I3Units.m, +2000 * I3Units.m],    # Set bounds to twice the size of the detector
                    BoundsY=[-2000 * I3Units.m, +2000 * I3Units.m],    # Set bounds to twice the size of the detector
                    BoundsZ=[-2000 * I3Units.m, +2000 * I3Units.m])    # Set bounds to twice the size of the detector

    # lilliput
    tray.AddService('I3GulliverMinuitFactory', 'Minuit',
                    Algorithm='SIMPLEX',    # Default
                    FlatnessCheck=True,     # Default
                    MaxIterations=1000,     # Only need 1000 iterations
                    MinuitPrintLevel=-2,    # Default
                    MinuitStrategy=2,       # Default
                    Tolerance=0.01)         # Set tolerance to 0.01

    # Seed the reduced SPESingle with the full SPESingle
    # lilliput
    tray.AddService('I3BasicSeedServiceFactory', 'MPESeed',
                    InputReadout=options['pulses_name'],
                    TimeShiftType='TFirst',
                    FirstGuesses=['MPEFit'])

    # Subset reconstruction time. This is slightly complicated. Each DOM is
    # placed into a partition based on (dom.string + dom.om) %
    # options['partitions']. So if a certain dom has (dom.string + dom.om) % 5
    # = 1 (right now options['partitions'] is 5), then that DOM is in the 1
    # partition. Now, when we do the fitter/pandel/reconstruction thing for
    # a particular partition, we want to do it on all the doms EXCEPT the ones
    # in that partition (this is that cross-validation). This is what
    # om_selection does. It lumps together all the TWSRTOfflinePulses for the
    # doms not in a particular partition into InIceRecoPulseSeriesPattern0...4. For
    # example, InIceRecoPulseSeriesPattern1 contains all the pulses except the
    # ones for the doms in the 1 partition. This is then fed into the
    # PandelFactory and the SimpleFitter.
    output_name = 'InIceRecoPulseSeriesPattern{}'

    tray.AddModule(om_partition, 'om_partition',
                   output_name=output_name,
                   options=options)

    for partition in range(options['partitions']):
        # lilliput
        tray.AddService('I3GulliverIPDFPandelFactory', 'MPEPandel{}'.format(partition),
                        InputReadout=output_name.format(partition),    # Use pulses given to thes function as arg
                        EventType='InfiniteMuon',                      # Default
                        Likelihood='MPE',                              # MPE
                        PEProb='GaussConvolutedFastApproximation',     # New approximation for convaluted
                        IceModel=2,                                    # Default
                        IceFile='',                                    # Default
                        AbsorptionLength=98.0 * I3Units.m,             # Default
                        JitterTime=4.0 * I3Units.ns,                   # Use small jitter time
                        NoiseProbability=10 * I3Units.hertz)           # Added a little noise term

        # gulliver-modules
        tray.AddModule('I3SimpleFitter', 'MPEFit{}'.format(partition),
                       # RandomService=SOBOL,                          # Name of randomizer service
                       SeedService='MPESeed',                          # Name of seed service
                       Parametrization='SimpleTrack',                  # Name of track parametrization service
                       LogLikelihood='MPEPandel{}'.format(partition),  # Name of likelihood service
                       Minimizer='Minuit')                             # Name of minimizer service

    # This uses the MPEFit's to calculate TotalCharge, RecoDistance, etc.
    tray.AddModule(dom_data, 'dom_data',
                   reco_fit='MPEFit{}',
                   options=options)

    # General

    # Calculate cut variables
    tray.AddSegment(direct_hits.I3DirectHitsCalculatorSegment, 'I3DirectHits',
                    PulseSeriesMapName=options['pulses_name'],
                    ParticleName='MPEFit',
                    OutputI3DirectHitsValuesBaseName='MPEFitDirectHits')

    tray.AddSegment(hit_multiplicity.I3HitMultiplicityCalculatorSegment, 'I3HitMultiplicity',
                    PulseSeriesMapName=options['pulses_name'],
                    OutputI3HitMultiplicityValuesName='HitMultiplicityValues')

    tray.AddSegment(hit_statistics.I3HitStatisticsCalculatorSegment, 'I3HitStatistics',
                    PulseSeriesMapName=options['pulses_name'],
                    OutputI3HitStatisticsValuesName='HitStatisticsValues')

    # Get ICAnalysisHits, DCAnalysisHits, and ICNHits
    tray.AddModule(count_hits, 'count_hits',
                   pulses_name=options['pulses_name'])

    # Move the cut variables into the top level of the frame.
    tray.AddModule(move_cut_variables, 'move_cut_variables',
                   direct_hits_name='MPEFitDirectHits',
                   fit_params_name='MPEFitFitParams')

    if args.sim:
        # Count the number of in ice muons and get the truth muon
        tray.AddModule(get_truth_muon, 'get_truth_muon')
        tray.AddModule(truth_endpoint, 'truth_endpoint')

    # Geoanalysis

    # Calculate the distance of each event to the detector border.
    tray.AddModule(calc_dist_to_border, 'calc_dist_to_border')

    # Write out the data to an I3 file
    tray.AddModule('I3Writer', 'I3Writer',
                   FileName=args.ofile,
                   SkipKeys=['InIceRecoPulseSeriesPattern.*'],
                   DropOrphanStreams=[icetray.I3Frame.DAQ],
                   Streams=[icetray.I3Frame.DAQ, icetray.I3Frame.Physics])

    tray.Execute()
    tray.Finish()

if __name__ == '__main__':
    main()
